File: 1461-count-all-valid-pickup-and-delivery-options/count-all-valid-pickup-and-delivery-options.py
class Solution:
    def countOrders(self, n: int) -> int:
        mod = int(1e9+7)
        @cache
        def solve(pick : int, deli : int) -> int:
            if pick == n and deli == 0: return 1
            ans = 0
            remPick = n - pick
            if remPick > 0:
                ans += remPick * solve(pick+1, deli+1)
            if deli > 0:
                ans += deli * solve(pick, deli-1)
            return ans % mod
        return solve(0, 0)

File: 1461-count-all-valid-pickup-and-delivery-options/count-all-valid-pickup-and-delivery-options.go
var mod = 1000000007

func solve(pick, del, total int, dp [][]int) int {
	if pick == total && del == 0 {
		return 1
	}
	if dp[pick][del] != -1 {
		return dp[pick][del]
	}

	var ans int
	remPick := total - pick

	if remPick > 0 {
		ans += remPick * solve(pick+1, del+1, total, dp)
		ans %= mod
	}

	if del > 0 {
		ans += del * solve(pick, del-1, total, dp)
		ans %= mod
	}

	dp[pick][del] = ans
	return ans
}

func countOrders(n int) int {
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
		for j := range dp[i] {
			dp[i][j] = -1
		}
	}
	return solve(0, 0, n, dp)
}